src/kernels/kernelFinder.ts

```diff
@@ -30,20 +36,20 @@ export class KernelFinder implements IKernelFinder {
     ): Promise<KernelConnectionMetadata[]> {
         this.startTimeForFetching = this.startTimeForFetching ?? new StopWatch();
 
-        const kernels: KernelConnectionMetadata[] = [];
+        // Wait all finders to warm up their cache first
+        await Promise.all(this._finders.map((finder) => finder.initialized));
 
-        const allKernels = await Promise.all(
-            this._finders.map((finder) => {
-                return finder.listContributedKernels(resource, cancelToken, useCache).then((kernels) => {
-                    this.finishListingKernels(kernels, useCache, finder.kind as 'local' | 'remote');
-                    return kernels;
-                });
-            })
-        );
+        if (cancelToken?.isCancellationRequested) {
+            return [];
+        }
 
-        allKernels.forEach((kernelList) => {
-            kernels.push(...kernelList);
-        });
+        const kernels: KernelConnectionMetadata[] = [];
+
+        for (const finder of this._finders) {
+            const contributedKernels = finder.listContributedKernels(resource);
```

Comment:
I know part of this is bolting the new push model onto the old method of working. But seems like now might be the PR to remove the useCache property from the kernelFinder api that the ControllerLoader is using. I think it's only use for telemetry at this point, and having the Loader call this function twice with a parameter that essentially does nothing sounds nice to get rid of.

Reply:
👍 with the new model, the telemetry might be no longer valid. Probably we can remove it and then add proper telemetry later.

Reply:
Yeah, this was the bit that I was wondering if we could get rid of at this point in the refactor process:
https://github.com/microsoft/vscode-jupyter/blob/f7428fb799dc912aaeaee66463f6f3b000484850/src/notebooks/controllers/controllerLoader.ts#L169
With your change hitting here I don't think this makes sense at all anymore.

Reply:
I pushed https://github.com/microsoft/vscode-jupyter/pull/11156/commits/3b205e9c5d42f9bbc88a14588117acaec455fcad to remove the cache but it might require more cleaning.